File: python/matrix.py
#	matrix

'''
	Class: player
	Description:
		representation of a player in a game
'''
class player:
	def __init__(self,Utility=None,Likelihood=None):
		self.Likelihood = Likelihood
		self.Utility = Utility
	def setUtility(self,Utility):
		self.Utility = Utility
	def getUtility(self):
		return self.Utility
	def setLikelihood(self,Likelihood):
		self.Likelihood = Likelihood	
	def getLikelihood(self):
		return self.Likelihood

'''
	Function: preferRational
	Description:
		determine the best course of action assuming pure rationality
	Input:
		Player (player): the player
	Output:
		return (Number): index of the chosen row of the matrix
'''
def preferRational(Player):
	WeightedUtility = evalRational(Player.getUtility(),Player.getLikelihood())
	return WeightedUtility.index(max(WeightedUtility))

'''
	Function: evalRational
	Description:
		evaluate the available options, assuming pure rationality
	Input:
		Utilities (List): the utility matrix
		Likelihood (List): list of weights representing how likely the player thinks each column of the matrix will occur
	Output:
		return (List): weighted utility matrix, taking into account both original utility values and likelihood
'''
def evalRational(Utilities,Likelihood):
	WeightedUtilities = []
	for Utility in Utilities:
		WeightedUtilities.append(weightedSum(Utility,Likelihood))
	return WeightedUtilities

'''
	Function: weightedSum
	Description:
		sum a list of integers, each integer multiplied by its corresponding value in a list of weights
	Input:
		Addends (List): list of integers
		Weights (List): list of weights
	Output:
		return (Number): the weighted sum
'''
def weightedSum(Addends,Weights):
	Sum = 0
	for A,Addend in enumerate(Addends):
		Sum += (Addend * Weights[A])
	return Sum

File: prolog/matrix.pl
%	matrix

:- [math].

/*
	Rule: setLikelihood
	Description:
		set the likelihood value for a player
	Input:
		Player (String): identifier for the player
		Likelihood (List): new value
*/
setLikelihood(Player,Likelihood) :-
	setProp(Player,'likelihood',Likelihood).

/*
	Rule: getLikelihood
	Description:
		get the likelihood value for a player
	Input:
		Player (String): identifier for the player
	Output:
		Likelihood (List): current value
*/
getLikelihood(Player,Likelihood) :-
	getProp(Player,'likelihood',Likelihood).

/*
	Rule: setUtility
	Description:
		set the utility value for a player
	Input:
		Player (String): identifier for the player
		Utility (List): new value
*/
setUtility(Player,Utility) :-
	setProp(Player,'utility',Utility).

/*
	Rule: getUtility
	Description:
		get the utility value for a player
	Input:
		Player (String): identifier for the player
	Output:
		Utility (List): current value
*/
getUtility(Player,Utility) :-
	getProp(Player,'utility',Utility).

/*
	Rule: setProp
	Description:
		helper for setting a property for a player
	Input:
		Player (String): identifier for the player
		Property (String): identifier for the property
		Value: new value
*/
setProp(Player,Property,Value) :-
	implode([Player,Property],'.',Name),
	nb_setval(Name,Value).

/*
	Rule: getProp
	Description:
		helper for getting a property for a player
	Input:
		Player (String): identifier for the player
		Property (String): identifier for the property
	Output:
		Value: current value
*/
getProp(Player,Property,Value) :-
	implode([Player,Property],'.',Name),
	nb_getval(Name,Value).

/*
	Rule: preferRational
	Description:
		determine the best course of action assuming pure rationality
	Input:
		Player (String): identifier for the player
	Output:
		Choice (Integer): index of the chosen row of the matrix
*/
preferRational(Player,Choice) :-
	getLikelihood(Player,Likelihood),
	getUtility(Player,Utility),
	evalRational(Utility,Likelihood,WeightedUtility),
	indexMax(WeightedUtility,Choice).

/*
	Rule: evalRational
	Description:
		evaluate the available options, assuming pure rationality
	Input:
		[] (List): the utility matrix
		Likelihood (List): list of weights representing how likely the player thinks each column of the matrix will occur
	Output:
		[] (List): weighted utility matrix, taking into account both original utility values and likelihood
*/
evalRational([U|Utility],Likelihood,[WU|WeightedUtility]) :-
	evalRational(Utility,Likelihood,WeightedUtility),
	weightedSum(U,Likelihood,Sum),
	WU is Sum.
	
evalRational([],_,[]).

/*
	Rule: weightedSum
	Description:
		sum a list of integers, each integer multiplied by its corresponding value in a list of weights
	Input:
		[] (List): list of integers
		[] (List): list of weights
	Output:
		Sum (Integer): the weighted sum
*/
weightedSum([A|Addends],[W|Weights],Sum) :-
	weightedSum(Addends,Weights,SumBelow),
	Sum is A * W + SumBelow.

weightedSum([],[],0).
